src/serde/de.rs

```diff
@@ -0,0 +1,883 @@
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_u16<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_u16(self.reader.read_i64().map(|x| x as u16)?);
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_u32<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_u32(self.reader.read_i64().map(|x| x as u32)?);
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_u64<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_u64(self.reader.read_i64().map(|x| x as u64)?);
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_f32<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_f32(self.reader.read_f32()?);
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_f64<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_f64(self.reader.read_f64()?);
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_char<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_char(
+            self.reader
+                .read_string()
+                .map(|x| x.to_string().chars().collect::<Vec<char>>()[0])?,
+        );
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_str<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_string(self.reader.read_string()?.to_string());
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_string<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_string(self.reader.read_str()?.to_owned());
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_bytes<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_bytes(self.reader.read_blob()?.as_slice());
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_byte_buf<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        let result = visitor.visit_byte_buf(self.reader.read_blob()?.as_slice().to_vec());
+        self.reader.next()?;
+        result
+    }
+
+    fn deserialize_option<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: Visitor<'de>,
+    {
+        if self.reader.is_null() {
+            self.reader.next()?;
```

Comment:
Yes, for `None` we have a null value so we juts do next here but for `Some` value we will go to the particualr type's visitor where it will do `read_next`